### DISCUSSION TITLE
NXP MIMXRT685 Support for "PowerQuad hardware accelerator" DSP?

### DISCUSSION THREAD

**Discussion by asteriskSF** [zephyrproject-rtos/zephyr#72195] [Q&A] [answered]
If I am developing on Zephyr with NXP MIMXRT685, am I able to utilized the DSP coprocessors?  

I currently have code that uses CMSIS DSP functions, will they take advantage of those coprocessors automatically?
Is there a dedicated library to use the coprocessors?
Do  I create a separate executable image for the coprocessors?
Any examples I can look at for guidance?
Are any of these features available in zephyr 3.5.0 or would I need to upgrade?

@DerekSnell maybe you have done this before or can point me to someone who has?

**Comment by DerekSnell** [comment, marked as answer]
Hi @asteriskSF , sorry for the delay.
In case this is still useful for you, or for others who find this discussion, here are some PowerQuad details for you:

> If I am developing on Zephyr with NXP MIMXRT685, am I able to utilized the DSP coprocessors?
Are any of these features available in zephyr 3.5.0 or would I need to upgrade?

You can use the PowerQuad coprocessor in a Zephyr application.  Currently there is no Zephyr abstraction or driver for the PowerQuad, but you can configure the PowerQuad in your app and call the CMSIS-DSP APIs from your app.  I am not aware today of a Zephyr example showing this.

Be aware, in addition to the PowerQuad coprocessor, the RT685 also includes a separate HiFi4 DSP core.  There is currently an open PR to enable this HiFi core, see https://github.com/zephyrproject-rtos/zephyr/pull/71061.

> I currently have code that uses CMSIS DSP functions, will they take advantage of those coprocessors automatically?
Is there a dedicated library to use the coprocessors?

Yes, you need the PowerQuad library included in the MCUXpresso SDK.  This has APIs compatible with CMSIS-DSP.  To learn more, see these application notes: [AN12383 ](https://www.nxp.com/docs/en/application-note/AN12383.pdf) and [AN12282](https://www.nxp.com/docs/en/application-note/AN12282.pdf). 

> Do I create a separate executable image for the coprocessors?

No, PowerQuad is a coprocessor for the Cortex-M33 core.  The PowerQuad will help accelerate the M33 image.

> Any examples I can look at for guidance?

You can refer to these [MCUXpresso SDK examples for the MIMXRT685-EVK board](https://github.com/nxp-mcuxpresso/mcux-sdk-examples/tree/main/evkmimxrt685/driver_examples/powerquad).

Best regards
